internal/middleware/compression.go

```diff
@@ -15,12 +15,59 @@
 package middleware
 
 import (
+	"bytes"
 	"compress/gzip"
 	"compress/zlib"
 	"io"
 	"net/http"
 )
 
+const (
+	headerContentEncoding = "Content-Encoding"
+	headerValueGZIP       = "gzip"
+)
+
+type CompressRoundTripper struct {
+	http.RoundTripper
+}
+
+func NewCompressRoundTripper(rt http.RoundTripper) *CompressRoundTripper {
+	return &CompressRoundTripper{
+		RoundTripper: rt,
+	}
+}
+
+func (r *CompressRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+
+	gzipWriter := gzip.NewWriter(nil)
+
+	// Gzip the body.
+	buf := bytes.NewBuffer([]byte{})
+	gzipWriter.Reset(buf)
+	_, copyErr := io.Copy(gzipWriter, req.Body)
+	req.Body.Close()
+
+	if err := gzipWriter.Close(); err != nil {
+		return nil, err
+	}
+
+	if copyErr != nil {
+		return nil, copyErr
+	}
+
+	// Create a new request since the docs say that we cannot modify the "req".
+	cReq, err := http.NewRequestWithContext(req.Context(), req.Method, req.URL.String(), buf)
+	if err != nil {
+		return nil, err
+	}
+
+	// Clone the headers and add gzip encoding header.
+	cReq.Header = req.Header.Clone()
+	cReq.Header.Add(headerContentEncoding, headerValueGZIP)
```

Comment:
Based on other places where we do in-place changes I think you should not worry to clone https://github.com/open-telemetry/opentelemetry-collector/blob/846b971758c92b833a9efaf742ec5b3e2fbd0c89/config/confighttp/confighttp.go#L91

Reply:
The docs say:
```
// RoundTrip should not modify the request, except for
// consuming and closing the Request's Body.
```